In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt

### PLAN

1. Pobranie danych
2. Przejrzeć co mamy
    - wielkosc data setu
    - jakie kolumny
    - jakiego typu dane
    - podejrzec probke
    - jesli dane sa we wlasciwym formacie: proste statystyki
3. Konwersja danych na odpowiedni format
4. Co z brakujacymi wartosciami?

#### Zaimportuj i przejrzyj dane

In [2]:
dane = pd.read_csv('data/DataWorkshop.csv')

In [3]:
print('Rozmiar danych: ', dane.shape)

Rozmiar danych:  (229929, 30)


In [4]:
print('Kolumny w datasecie: ')
dane.columns

Kolumny w datasecie: 


Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków'],
      dtype='object')

In [5]:
print('Typy danych: ')
dane.dtypes

Typy danych: 


Data zdarzenia                        object
Godzina                               object
GPS x                                 object
GPS y                                 object
Miejscowość                           object
Ulica                                 object
Numer domu                            object
Numer drogi                           object
KM HM (Pikietaż)                      object
Obszar                                object
Skrzyżowanie                          object
Geometria drogi                       object
Charakterystyka miejsca zdarzenia     object
Oświetlenie                           object
Warunki atmosferyczne                 object
Rodzaj zdarzenia                      object
Wina                                  object
Rodzaj drogi                          object
Nawierzchnia drogi                    object
Stan nawierzchni                      object
Sygnalizacja świetlna                 object
Oznakowanie poziome                   object
Liczba pas

In [6]:
print('Brakujace wartosci: ')

missing = dane.isnull().sum()
print(missing)

Brakujace wartosci: 
Data zdarzenia                            0
Godzina                                   0
GPS x                                     0
GPS y                                     0
Miejscowość                               0
Ulica                                 20728
Numer domu                            93884
Numer drogi                               0
KM HM (Pikietaż)                     194888
Obszar                                    1
Skrzyżowanie                         166779
Geometria drogi                       66008
Charakterystyka miejsca zdarzenia         2
Oświetlenie                               0
Warunki atmosferyczne                     1
Rodzaj zdarzenia                          0
Wina                                      0
Rodzaj drogi                              1
Nawierzchnia drogi                        1
Stan nawierzchni                          1
Sygnalizacja świetlna                     0
Oznakowanie poziome                       2
Liczba pasó

In [7]:
print('Brakujace wartosci jako procent zbioru, %: ')
index=[]
lst=[]
for item in missing.iteritems():
    index.append(item[0])
    lst.append(round((item[1]/dane.shape[0])*100,4))
missing_percentage=pd.Series(lst, index=index)
print(missing_percentage)

Brakujace wartosci jako procent zbioru, %: 
Data zdarzenia                        0.0000
Godzina                               0.0000
GPS x                                 0.0000
GPS y                                 0.0000
Miejscowość                           0.0000
Ulica                                 9.0150
Numer domu                           40.8317
Numer drogi                           0.0000
KM HM (Pikietaż)                     84.7601
Obszar                                0.0004
Skrzyżowanie                         72.5350
Geometria drogi                      28.7080
Charakterystyka miejsca zdarzenia     0.0009
Oświetlenie                           0.0000
Warunki atmosferyczne                 0.0004
Rodzaj zdarzenia                      0.0000
Wina                                  0.0000
Rodzaj drogi                          0.0004
Nawierzchnia drogi                    0.0004
Stan nawierzchni                      0.0004
Sygnalizacja świetlna                 0.0000
Oznakowanie

In [8]:
# Probka danych
dane.sample(50)

Data zdarzenia   Godzina      GPS x      GPS y       Miejscowość  \
71102      2017-05-05  09:30:00  18*37'480  50*23'548         PYSKOWICE   
139222     2018-07-07  18:30:00  19*08'233  50*17'384         SOSNOWIEC   
19460      2016-05-25  12:30:00  18*53'039  50*10'451           MIKOŁÓW   
177168     2019-02-21  09:30:00  19*08'062  50*47'395       CZĘSTOCHOWA   
96007      2017-10-08  00:05:00  19*19'411  50*44'351           ZRĘBICE   
92758      2017-09-21  14:35:00  19*01'005  50*15'079          KATOWICE   
23819      2016-06-25  08:10:00  19*10'041  50*12'060         MYSŁOWICE   
93491      2017-09-22  22:10:00  18*59'301  50*07'582             TYCHY   
52999      2016-12-30  17:45:00  18*13'286  50*05'448          RACIBÓRZ   
223488     2019-11-26  07:45:00  18*55'189  50*26'157     NAKŁO ŚLĄSKIE   
142422     2018-07-24  08:00:00  19*08'325  50*16'057         SOSNOWIEC   
185529     2019-04-14  20:55:00  18*59'395  50*38'256         KAMIENICA   
212290     2019-09-21  11:59:00  19*06'266  50*19'102           CZELADŹ   
223982     2019-11-29  08:15:00  19*02'502  49*48'457     BIELSKO-BIAŁA   
123940     2018-03-29  10:24:00  19*15'060  50*13'022          JAWORZNO   
125439     2018-04-09  17:05:00  19*00'218  50*14'519          KATOWICE   
21228      2016-06-06  19:30:00  18*12'312  50*05'501          RACIBÓRZ   
103169     2017-11-17  21:02:00  19*39'368  50*52'138            CUDKÓW   
22113      2016-06-13  06:55:00  18*42'514  50*01'122              ŻORY   
122407     2018-03-19  10:00:00  19*09'562  50*19'301  DĄBROWA GÓRNICZA   
16148      2016-05-04  11:30:00  18*27'179  50*02'485            RADLIN   
160857     2018-11-19  13:16:00  19*11'595  50*07'289      CHEŁM ŚLĄSKI   
2009       2016-01-19  11:10:00  18*25'563  50*05'499             PIECE   
200750     2019-07-15  10:40:00  19*03'310  49*49'280     BIELSKO-BIAŁA   
227485     2019-12-17  18:25:00  18*47'043  50*18'316            ZABRZE   
77398      2017-06-13  15:55:00  19*07'116  50*21'222            BĘDZIN   
199250     2019-07-05  06:45:00  19*05'314  49*29'568     RYCERKA DOLNA   
204536     2019-08-08  14:20:00  19*10'492  50*19'384  DĄBROWA GÓRNICZA   
76261      2017-06-06  22:15:00  19*01'405  50*06'041             TYCHY   
32852      2016-08-27  11:30:00  19*09'078  50*19'229            BĘDZIN   
29743      2016-08-07  23:10:00  19*00'427  50*15'210          KATOWICE   
44982      2016-11-13  16:46:00  19*08'044  50*05'013            BIERUŃ   
94850      2017-10-02  16:30:00  18*54'554  50*20'366             BYTOM   
105714     2017-12-01  18:06:00  18*57'124  50*17'503           CHORZÓW   
30298      2016-08-07  21:00:00  18*54'052  50*21'000             BYTOM   
138831     2018-07-04  17:40:00  18*30'387  50*03'360            RYBNIK   
46977      2016-11-25  16:30:00  18*41'227  50*02'191              ŻORY   
44890      2016-11-10  17:20:00  18*27'167  50*00'527  WODZISŁAW ŚLĄSKI   
177069     2019-02-19  17:00:00  18*39'258  50*15'430           GLIWICE   
122503     2018-03-19  10:40:00  18*31'241  50*06'061            RYBNIK   
98059      2017-10-19  16:25:00  18*38'334  50*18'096           GLIWICE   
55568      2017-01-17  16:15:00  18*59'219  50*13'192          KATOWICE   
22660      2016-06-16  14:50:00  19*00'004  50*36'270            LUBSZA   
44398      2016-11-08  21:40:00  19*06'094  50*48'150       CZĘSTOCHOWA   
99556      2017-10-27  19:18:00  19*03'137  50*14'329          KATOWICE   
21261      2016-06-08  11:30:00  18*58'284  50*07'114             TYCHY   
54308      2017-01-09  15:20:00  19*10'271  50*17'293         SOSNOWIEC   
30705      2016-08-09  16:30:00  18*47'532  49*47'302           SKOCZÓW   
145396     2018-08-17  14:30:00  19*01'596  49*49'029     BIELSKO-BIAŁA   
197927     2019-06-27  17:08:00  19*49'354  50*37'230       SZCZEKOCINY   

                              Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
71102                  WYSZYŃSKIEGO        NaN      P2905S              NaN   
139222                 

### Konwersja typu danych

#### Data i czas

In [9]:
# Utworzenie kopii zbioru danych. Dalsza konwersja odbywac sie bedzie na kopii danych
df=dane.copy()

In [10]:
# Data zdarzenia i Godzina scalone do jednej kolumny
print('kolumna DATA ZDARZENIA: ', df['Data zdarzenia'].dtypes)
print('kolumna GODZINA: ', df['Data zdarzenia'].dtypes)
df['data_full'] = df['Data zdarzenia'] + ' ' + df['Godzina']
df.sample(10)

kolumna DATA ZDARZENIA:  object
kolumna GODZINA:  object


Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
149564     2018-09-13  10:15:00  18*48'230  50*22'370                ZABRZE   
138501     2018-07-03  09:00:00  19*01'513  50*19'327  SIEMIANOWICE ŚLĄSKIE   
96060      2017-10-08  13:31:00  19*23'575  50*40'340           ZŁOTY POTOK   
74894      2017-05-30  00:56:00  19*06'198  50*50'155           CZĘSTOCHOWA   
159419     2018-11-08  08:40:00  18*25'232  49*59'313             CZYŻOWICE   
62960      2017-03-10  12:20:00  19*25'362  50*28'519             ZAWIERCIE   
121602     2018-03-13  15:15:00  19*00'594  50*22'011             WOJKOWICE   
58238      2017-02-06  09:17:00  18*45'366  49*59'439                SUSZEC   
104783     2017-11-27  17:55:00  18*55'309  50*16'241        ŚWIĘTOCHŁOWICE   
151747     2018-09-25  16:23:00  19*08'225  50*18'033             SOSNOWIEC   

                              Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
149564         LUDWIKI WAWRZYŃSKIEJ          1    G140005S              NaN   
138501             HENRYKA KRUPANKA         83          94              NaN   
96060                           NaN        NaN        W793              NaN   
74894                        ŁÓDZKA        243      P6601S              NaN   
159419                         NOWA         51      P5017S              NaN   
62960   GEN. WŁADYSŁAWA SIKORSKIEGO          6      P1701S              NaN   
121602                  SOBIESKIEGO        368          I2              NaN   
58238                    WYZWOLENIA         46      P4100S              NaN   
104783                       LIPOWA         52      GMINNA              NaN   
151747                          NaN        NaN          94              NaN   

                      Obszar  ... Oznakowanie poziome  \
149564     Obszar zabudowany  ...              Nie ma   
138501     Obszar zabudowany  ...                Jest   
96060   Obszar niezabudowany  ...                Jest   
74894      Obszar zabudowany  ...              Nie ma   
159419     Obszar zabudowany  ...                Jest   
62960      Obszar zabudowany  ...                Jest   
121602     Obszar zabudowany  ...              Nie ma   
58238      Obszar zabudowany  ...                Jest   
104783     Obszar zabudowany  ...              Nie ma   
151747  Obszar niezabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
149564                      2.0                        Nieprawidłowe cofanie   
138501                      2.0                         Zmęczenie, zaśnięcie   
96060                       NaN  Niedostosowanie prędkości do warunków ruchu   
74894                       NaN                                          NaN   
159419                      2.0  Niezachowanie bezp. odleg. między pojazdami   
62960                       NaN        Nieustąpienie pierwszeństwa przejazdu   
121602                      2.0                        Nieprawidłowe cofanie   
58238                       NaN                      Nieprawidłowe skręcanie   
104783                      NaN                        Nieprawidłowe cofanie   
151747                      2.0                               Inne przyczyny   

       Zachowanie pieszego Inne przyczyny Liczba rannych w zdarzeniu  \
149564                 NaN            NaN                          0   
138501                 NaN            NaN                          0   
96060                  NaN            NaN                          2   
74894                  NaN    Stan jezdni                          0   
159419                 NaN            NaN                          0   
62960                  NaN            NaN                          0   
121602                 NaN            NaN                          0   
58238                  NaN            NaN                          0   
104783                 NaN            NaN                          0   
151747                 NaN            

In [11]:
# Data zdarzenia i Godzina do typu datetime
df['data_full'] = pd.to_datetime(df['data_full'], format='%Y-%m-%d %H:%M:%S')
df.sample(10)
# TO-DO: Jaki to dzien tygodnia? Jaki to dzien miesiaca? To dzien czy noc? Jaka faza ksiezyca itd..

Data zdarzenia   Godzina      GPS x      GPS y         Miejscowość  \
9568       2016-03-08  12:00:00  18*59'359  50*14'438            KATOWICE   
135471     2018-06-12  18:46:00  19*04'412  50*14'233            KATOWICE   
171955     2019-01-19  10:20:00  19*08'099  50*18'240              BĘDZIN   
5835       2016-02-17  13:00:00  18*58'400  50*12'320            KATOWICE   
57938      2017-02-01  15:05:00  18*59'032  49*48'311       BIELSKO-BIAŁA   
131728     2018-05-18  10:15:00  19*00'314  50*47'050             WYRAZÓW   
1342       2016-01-14  09:30:00  19*00'580  50*15'390            KATOWICE   
114415     2018-01-24  16:10:00  19*06'327  50*18'592              BĘDZIN   
50693      2016-12-16  14:29:00  19*28'208  50*28'399           ZAWIERCIE   
205435     2019-08-14  07:40:00  18*55'302  50*29'414  MIASTECZKO ŚLĄSKIE   

                         Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
9568                      POLA         32      P6501S              NaN   
135471             SZOPIENICKA         43      P6485S              NaN   
171955             27 STYCZNIA        NaN      GMINNA              NaN   
5835      GEN.ZYGMUNTA JANKEGO        150      P6442S              NaN   
57938                  TWÓRCZA         11    G160901S              NaN   
131728           CZĘSTOCHOWSKA        NaN          46              NaN   
1342               SŁOWACKIEGO         12      P6518S              NaN   
114415                     NaN        NaN          86             15,1   
50693   IGNACEGO PADEREWSKIEGO        NaN          78              NaN   
205435                WOŹNICKA        NaN        W908              NaN   

                      Obszar  ... Oznakowanie poziome  \
9568       Obszar zabudowany  ...                Jest   
135471     Obszar zabudowany  ...                Jest   
171955     Obszar zabudowany  ...              Nie ma   
5835       Obszar zabudowany  ...                Jest   
57938      Obszar zabudowany  ...                Jest   
131728     Obszar zabudowany  ...                Jest   
1342       Obszar zabudowany  ...                Jest   
114415  Obszar niezabudowany  ...                Jest   
50693      Obszar zabudowany  ...                Jest   
205435     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
9568                        NaN                       Nieprawidłowe: cofanie   
135471                      2.0        Nieustąpienie pierwszeństwa przejazdu   
171955                      2.0                       Nieprawidłowe omijanie   
5835                        NaN    Niezachowanie bezp. odl. między pojazdami   
57938                       NaN                        Nieprawidłowe cofanie   
131728                      2.0                        Nieprawidłowe cofanie   
1342                        NaN                  Inne przyczyny (od 11.2015)   
114415                      4.0          Nieprawidłowe zmienianie pasa ruchu   
50693                       NaN        Nieustąpienie pierwszeństwa przejazdu   
205435                      2.0  Niedostosowanie prędkości do warunków ruchu   

       Zachowanie pieszego Inne przyczyny Liczba rannych w zdarzeniu  \
9568                   NaN            NaN                          0   
135471                 NaN            NaN                          0   
171955                 NaN            NaN                          0   
5835                   NaN            NaN                          0   
57938                  NaN            NaN                          0   
131728                 NaN            NaN                          0   
1342                   NaN            NaN                          0   
114415                 NaN            NaN                          0   
50693                  NaN            NaN                          0   
205435                 NaN            NaN                          1   

       Liczba zabitych w zdarzeniu Liczba 

#### Wspolrzedne geograficzne

Domyslny zapis wspolrzednych to D\*M'S([D]egree, [M]inute, [S]cond) np. 18\*53'289. Jednak, gdy pierwsza cyfra sekund jest zero, czasem wystepuje zapis D\*M''S np. 18\*53''11. Wystepowaly takze bledne zapisy takie jak 4\*\*56'569 i 50\*09'\*01.

In [12]:
def dmsToDecimal(coordinate_as_string):
    try:
        # Zapis domyslny to D*M'S([D]egree, [M]inute, [S]cond) zeby sprawnie podzielic kazda wspolrzedna na liste trzyelementowa
        # zamieniam ' na *, a nastepnie dziele orzymany rezultat znakami *. W trakcie konwersji danych napotkalam rozne nieprawidlowe
        # zapisy opisane w komorce powyzej. Wszystkie wystepujace nietypowe znaki takie jak **, '* i '' zamieniam na * i dziele znakiem *
        degree, minute, second = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
    except:
        # Jesli nie dzieli na liste trzyelementowa to najprawdopodobniej trzeci element jest pusty(lista dwuelementowa).
        # Wtedy pierwszy element to stopien, drugi minuta. Sekundy zas sa rowne 0.
        degree, minute = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
        second=0
    # W niektorych wypadkach udaje sie podzielic wsadowego stringa tak aby uzyskac ostatnia czesc jako pusty string.
    #Ponizej przeksztalcam pustego stringa w wartosc 0
    if second == "" or second == " ":
        second=0
    # Wlasciwe przeksztalcenie wspolrzednej z wartosci katowych na wartosc liczbowa
    coordinate_as_float = float(degree) + float(minute)/60 + float(second)/(60*60)
    return coordinate_as_float
# Przygotowana wyzej funkcje wywoluje na calych kolumnach dancych
df['GPS x'] = df['GPS x'].apply(lambda x: dmsToDecimal(x))
df['GPS y'] = df['GPS y'].apply(lambda x: dmsToDecimal(x))
df.sample(10)
# TO-DO Wyszukac i wykorzystac biblioteke okreslajaca wspolrzedne GPS na podstawie adresu

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
126797     2018-04-17  16:15:00  19.065556  50.584722             OŻAROWICE   
220023     2019-11-04  20:45:00  19.171944  50.821111           CZĘSTOCHOWA   
178190     2019-02-28  18:30:00  18.692778  50.303889               GLIWICE   
10005      2016-03-17  09:50:00  18.719444  49.766667               CIESZYN   
123957     2018-03-29  10:40:00  19.130556  50.319722  SIEMIANOWICE ŚLĄSKIE   
184027     2019-04-05  12:15:00  19.007222  50.319444              KATOWICE   
79784      2017-06-29  13:35:00  18.859444  50.589722               POŁOMIA   
95342      2017-10-05  06:30:00  19.222778  50.317778             MYSŁOWICE   
28844      2016-07-30  05:30:00  18.608611  50.012222      JASTRZĘBIE-ZDRÓJ   
98413      2017-10-22  13:25:00  18.731389  50.351111               GLIWICE   

                        Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
126797               TOPOLOWA        NaN      P3236S              NaN   
220023           ALEJA POKOJU         14      P6628S              NaN   
178190               RYBNICKA        207          I2              NaN   
10005                 BIELSKA        100      P2619S              NaN   
123957         KS.JANA KAPICY         9A      P9620S              NaN   
184027  ZBIGNIEWA CYBULSKIEGO          3    G100308S              NaN   
79784              PYSKOWICKA        NaN      P2905S              NaN   
95342                  3 MAJA          7      P8800S              NaN   
28844                GÓRNICZA          1    G220129S              NaN   
98413                 ŁUŻYCKA         12      P7211S              NaN   

                      Obszar  ... Oznakowanie poziome  \
126797     Obszar zabudowany  ...              Nie ma   
220023     Obszar zabudowany  ...              Nie ma   
178190     Obszar zabudowany  ...                Jest   
10005      Obszar zabudowany  ...                Jest   
123957     Obszar zabudowany  ...                Jest   
184027     Obszar zabudowany  ...                Jest   
79784   Obszar niezabudowany  ...              Nie ma   
95342      Obszar zabudowany  ...                Jest   
28844      Obszar zabudowany  ...                Jest   
98413      Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
126797                      2.0        Nieustąpienie pierwszeństwa przejazdu   
220023                      2.0                        Nieprawidłowe cofanie   
178190                      2.0        Nieustąpienie pierwszeństwa przejazdu   
10005                       2.0        Nieudzielenie pierwszeństwa przejazdu   
123957                      3.0        Nieustąpienie pierwszeństwa przejazdu   
184027                      2.0  Niezachowanie bezp. odleg. między pojazdami   
79784                       2.0                                          NaN   
95342                       NaN                                          NaN   
28844                       NaN  Niezachowanie bezp. odleg. między pojazdami   
98413                       2.0        Nieustąpienie pierwszeństwa przejazdu   

       Zachowanie pieszego                Inne przyczyny  \
126797                 NaN                           NaN   
220023                 NaN                           NaN   
178190                 NaN                           NaN   
10005                  NaN                           NaN   
123957                 NaN                           NaN   
184027                 NaN                           NaN   
79784                  NaN                   Stan jezdni   
95342                  NaN  Obiekty, zwierzęta na drodze   
28844                  NaN                           NaN   
98413                  NaN                           NaN   

       Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu Liczba kolizji  \
126797                          0                           0              1   
220

#### Miejsce zdarzenia



In [13]:
# Zmiana nazwy kolumn na nazwy bez spacji i polskich znakow
# df.rename(columns={'Miejscowość': 'Miejscowosc', 'Numer domu': "NrDomu", "Numer drogi": "NrDrogi"}, inplace=True)
df.columns

Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków',
       'data_full'],
      dtype='object')

In [14]:
# Pikietaz okresla miejsce zdarzenia na drodze poza terenem zabudowanym
#Zapisywany jako xx,yyy gdzie:
#xx – całkowita liczba kilometrów od początku drogi;
#yyy – całkowita liczba metrów liczona od ostatniego pełnego kilometra.
# W zbiorze danych prawdopodobnie jako wartosc w km, tyle ze z uzyciem "," jako separatora dziesietnego
def pikietazToFloat(pikietazRaw):
    if isinstance(pikietazRaw, str):
        pikietazAsFloat = float(pikietazRaw.replace(",", "."))
    else:
        pikietazAsFloat = pikietazRaw
    return pikietazAsFloat
df['KM HM (Pikietaż)']=df['KM HM (Pikietaż)'].apply(lambda x: pikietazToFloat(x))
# df.rename(columns={'KM HM (Pikietaż)': 'KM_HM_Pikietaz'})
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y       Miejscowość  \
108204     2017-12-15  20:24:39  18.824167  50.363056           GLIWICE   
198625     2019-07-02  12:20:00  19.250278  50.385000         MYSŁOWICE   
1058       2016-01-12  11:25:00  18.266111  50.114167         CYPRZANÓW   
181289     2019-03-19  06:51:00  19.380278  50.477778  DĄBROWA GÓRNICZA   
125340     2018-04-09  07:39:00  19.047222  50.353056          KATOWICE   
115893     2018-02-04  09:05:00  18.818056  50.489722   TARNOWSKIE GÓRY   
15092      2016-04-26  07:50:00  19.074444  50.290000          KATOWICE   
215661     2019-10-13  15:00:00  19.039167  49.815556           SZCZYRK   
145140     2018-08-15  22:15:00  19.120278  50.368611          KATOWICE   
187111     2019-04-24  14:50:00  18.798889  50.393611           GLIWICE   

                           Ulica Numer domu Numer drogi  KM HM (Pikietaż)  \
108204               BŁ.CZESŁAWA         60      P7222S               NaN   
198625                 KATOWICKA         75          I2               NaN   
1058                    JANOWSKA         25      P3505S               NaN   
181289                   SZOSOWA         70        W796               NaN   
125340                       NaN        NaN          A4             339.0   
115893                  GLIWICKA        NaN          78              72.5   
15092                 PANEWNICKA        NaN      P6494S               NaN   
215661                 MYŚLIWSKA         43        W942               NaN   
145140  WALENTEGO ROŹDZIEŃSKIEGO        NaN          79               NaN   
187111    STANISŁAWA KONARSKIEGO         25    G130328S               NaN   

                      Obszar  ... Oznakowanie poziome  \
108204     Obszar zabudowany  ...                Jest   
198625     Obszar zabudowany  ...              Nie ma   
1058       Obszar zabudowany  ...                Jest   
181289     Obszar zabudowany  ...                Jest   
125340  Obszar niezabudowany  ...                Jest   
115893  Obszar niezabudowany  ...                Jest   
15092      Obszar zabudowany  ...                Jest   
215661     Obszar zabudowany  ...              Nie ma   
145140     Obszar zabudowany  ...                Jest   
187111     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
108204                      2.0                                          NaN   
198625                      2.0                               Inne przyczyny   
1058                        NaN                                          NaN   
181289                      2.0  Niezachowanie bezp. odleg. między pojazdami   
125340                      8.0  Niezachowanie bezp. odleg. między pojazdami   
115893                      2.0  Niedostosowanie prędkości do warunków ruchu   
15092                       NaN        Nieudzielenie pierwszeństwa przejazdu   
215661                      2.0  Niezachowanie bezp. odleg. między pojazdami   
145140                      5.0          Nieprawidłowe zmienianie pasa ruchu   
187111                      3.0  Niezachowanie bezp. odleg. między pojazdami   

                                      Zachowanie pieszego  \
108204                                                NaN   
198625                                                NaN   
1058    Nieostrożne wejście na jezdnię: przed jadącym ...   
181289                                                NaN   
125340                                                NaN   
115893                                                NaN   
15092                                                 NaN   
215661                                                NaN   
145140                                                NaN   
187111                                                NaN   

                      Inne przyczyny Liczba rannych w zdarzeniu  \
108204  Obiekty, zwierzęta na drodze                          0   
198625            

In [15]:
# Obszar: zabudowany/niezabudowany
# Przypisane zostaly wartosci logiczne: Obszar zabudowany-True, Obszar niezabudowany False.
# W razie potrzeby takie wartosci mozna wygodnie przeksztalcic na int o wartosciach odpowiednio 1 i 0
df['Obszar'].value_counts()
df['Obszar'] = df['Obszar'].apply(lambda x: True if x=='Obszar zabudowany' else False)
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y  Miejscowość  \
178204     2019-02-27  16:40:00  18.775278  50.076389         ŻORY   
7653       2016-02-29  21:30:00  19.198889  50.253889     KATOWICE   
161185     2018-11-20  15:00:00  19.180556  50.895278  CZĘSTOCHOWA   
207267     2019-08-26  07:00:00  19.062778  50.265833     KATOWICE   
16089      2016-04-28  08:00:00  19.225556  50.965278  CZĘSTOCHOWA   
103616     2017-11-21  13:40:00  18.643889  50.396389      GLIWICE   
107877     2017-12-14  15:10:00  19.205833  50.344722    MYSŁOWICE   
87010      2017-08-19  08:43:00  19.667222  50.792778     PODLESIE   
155972     2018-10-19  17:35:00  19.113889  50.284167     KATOWICE   
23916      2016-06-22  17:55:00  18.729167  50.369167      ŻERNICA   

                           Ulica Numer domu Numer drogi  KM HM (Pikietaż)  \
178204                MIKOŁOWSKA         13      P9004S               NaN   
7653                 GOSPODARCZA        NaN      P6433S               NaN   
161185               DEKABRYSTÓW        NaN      P6611S               NaN   
207267                MIKOŁOWSKA         52      P6479S               NaN   
16089            RYNEK WIELUŃSKI        NaN    G110744S               NaN   
103616              STAROMIEJSKA        NaN      P7211S               NaN   
107877                     KACZA          4    G240006S               NaN   
87010                        NaN         31      P1098S               NaN   
155972  WALENTEGO ROŹDZIEŃSKIEGO         93          86               NaN   
23916           JÓZEFA SZAFRANKA        NaN      P2916S               NaN   

        Obszar  ... Oznakowanie poziome Liczba pasów w zdarzeniu  \
178204    True  ...                Jest                      2.0   
7653      True  ...                Jest                      NaN   
161185    True  ...                Jest                      4.0   
207267    True  ...                Jest                      3.0   
16089     True  ...                Jest                      NaN   
103616    True  ...                Jest                      2.0   
107877    True  ...              Nie ma                      NaN   
87010     True  ...                Jest                      NaN   
155972    True  ...                Jest                      4.0   
23916     True  ...                Jest                      2.0   

                             Zachowanie kierującego Zachowanie pieszego  \
178204                                          NaN                 NaN   
7653          Nieudzielenie pierwszeństwa przejazdu                 NaN   
161185        Nieustąpienie pierwszeństwa przejazdu                 NaN   
207267                                          NaN                 NaN   
16089         Nieudzielenie pierwszeństwa przejazdu                 NaN   
103616                      Nieprawidłowe skręcanie                 NaN   
107877                        Nieprawidłowe cofanie                 NaN   
87010   Niezachowanie bezp. odleg. między pojazdami                 NaN   
155972        Nieustąpienie pierwszeństwa przejazdu                 NaN   
23916         Nieustąpienie pierwszeństwa przejazdu                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
178204    Stan jezdni                          0                           0   
7653              NaN                          0                           0   
161185            NaN                          0                           0   
207267    Nieustalone                          0                           0   
16089             NaN                          0                           0   
103616            NaN                          0                           0   
107877            NaN                          0                           0   
87010             NaN                          0                           0   
155972            NaN                          0                           0   
23916        

In [16]:
# Skrzyzowanie
# Zmiana nazwy na niezawierajaca polskich znakow
df['Skrzyżowanie'].unique()
df['Skrzyżowanie'].value_counts()
# Zamiana wartosci tekstowych na int zgodnie z zasada:
# 0 : O ruchu okrężnym; 1 : Równorzędne; 2 : Z drogą z pierwsz.
skrzyzowaniemap = {'O ruchu okrężnym':0, 'Równorzędne':1, 'Z drogą z pierwsz.':2}
df=df.applymap(lambda x: skrzyzowaniemap.get(x) if x in skrzyzowaniemap else x)
# df.rename(columns={'Skrzyżowanie': 'Skrzyzowanie'})

In [17]:
df['Geometria drogi'].value_counts()
# Geometria pozioma: 
# 0: Odcinek prosty; 1: Zakret(luk)

# Geometria pionowa: 
# 0: Spadek; 1: wzniesienie; 2: wierzcholek wzniesienia



Odcinek prosty                      142703
Zakręt, łuk                          16917
Spadek                                1866
Wzniesienie                           1317
Odcinek prosty; Spadek                 430
Odcinek prosty; Wzniesienie            329
Zakręt, łuk; Spadek                    179
Zakręt, łuk; Wzniesienie               108
Wierzchołek wzn.                        50
Odcinek prosty; Wierzchołek wzn.        16
Zakręt, łuk; Wierzchołek wzn.            5
Odcinek prosty; Zakręt, łuk              1
Name: Geometria drogi, dtype: int64

In [27]:
#df['Geometria_drogi_poziom'] = df['Geometria drogi'].copy()
#df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 0 if 'Odcinek prosty' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'Zakręt' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'łuk' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane czy droga byla prosta czy byl to zakret?
df['Geometria_drogi_poziom'].value_counts()

0                   143479
1                    17209
Spadek                1866
Wzniesienie           1317
Wierzchołek wzn.        50
Name: Geometria_drogi_poziom, dtype: int64

In [28]:
df['Geometria_drogi_pion'] = df['Geometria drogi'].copy()
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 0 if 'Spadek' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 1 if 'Wzniesienie' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 2 if 'Wierzchołek wzn.' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane uksztaltowanie terenu?
df['Geometria_drogi_pion'].value_counts()

Odcinek prosty                 142703
Zakręt, łuk                     16917
0                                2475
1                                1754
2                                  71
Odcinek prosty; Zakręt, łuk         1
Name: Geometria_drogi_pion, dtype: int64